[DDP] Support not all outputs used in loss calculation

rohan-varma opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#57081 [DDP] Support not all outputs used in loss calculation**

Changes in this diff:
1. Enable passthrough autograd function when find_unused_parameters=True.
2. With above, move `prepare_for_backward` which does unused parameter checking logic to beginning of backwards pass, only when `find_unused_parameters=True`. 
3. Enhance process of unused parameter checking to account for outputs not being used in loss.

The way (3) is implemented is by triggering the autograd hook corresponding to parameters that did not participate in loss computation. Since they did not participate, the autograd hook is triggered with a gradient of None, and the reducer handles this appropriately to ensure that the gradient is not touched.

Tested by ensuring that when a model output is not used in loss, the corresponding grad is not modified. Also verified that the grads are the same in local vs DDP training case. Also verified that gradients are not touched in this case, i.e. if grad is originally `None`, it stays as `None`, not zero, after.

Note that in this diff we are not enabling the pass through autograd function for regular case find_unused_parameters=False because that has a much bigger blast radius and needs additional careful analysis especially with regard to the performance. 


Differential Revision: [D28048628](https://our.internmc.facebook.com/intern/diff/D28048628/)

rohan-varma committed:
[DDP] Enable passthrough for all cases

Enabling passthrough autograd function for all cases, not just static
graph. Existing unittests should ensure correctness.

Differential Revision: [D28048628](https://our.internmc.facebook.com/intern/diff/D28048628/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[DDP] Enable passthrough for all cases"


Enabling passthrough autograd function for all cases, not just static
graph. Existing unittests should ensure correctness.

Differential Revision: [D28048628](https://our.internmc.facebook.com/intern/diff/D28048628/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[DDP] Support not all outputs used in loss calculation"


Changes in this diff:
1. Enable passthrough autograd function for all cases, not just static.
2. With above, move `prepare_for_backward` which does unused parameter checking logic to beginning of backwards pass
3. Enhance process of unused parameter checking to account for outputs not being used in loss.

The way (3) is implemented is by looking at the gradient outputs at the beginning of backwards. If the gradient output corresponding to a model output is zero it means that it was not involved in loss computation (it could also mean that the gradient was indeed all zeros, but this is practically equivalent). In this case, we get the corresponding model output and mark that parameter as unused (which is equivalent to a zero gradient being accumulated).

Tested by ensuring that when a model output is not used in loss, the corresponding grad is not modified. Also verified that the grads are the same in local vs DDP training case.

Differential Revision: [D28048628](https://our.internmc.facebook.com/intern/diff/D28048628/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[DDP] Support not all outputs used in loss calculation"


Changes in this diff:
1. Enable passthrough autograd function for all cases, not just static.
2. With above, move `prepare_for_backward` which does unused parameter checking logic to beginning of backwards pass
3. Enhance process of unused parameter checking to account for outputs not being used in loss.

The way (3) is implemented is by looking at the gradient outputs at the beginning of backwards. If the gradient output corresponding to a model output is zero it means that it was not involved in loss computation (it could also mean that the gradient was indeed all zeros, but this is practically equivalent). In this case, we get the corresponding model output and mark that parameter as unused (which is equivalent to a zero gradient being accumulated).

Tested by ensuring that when a model output is not used in loss, the corresponding grad is not modified. Also verified that the grads are the same in local vs DDP training case.

Differential Revision: [D28048628](https://our.internmc.facebook.com/intern/diff/D28048628/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[DDP] Support not all outputs used in loss calculation"


Changes in this diff:
1. Enable passthrough autograd function for all cases, not just static.
2. With above, move `prepare_for_backward` which does unused parameter checking logic to beginning of backwards pass
3. Enhance process of unused parameter checking to account for outputs not being used in loss.

The way (3) is implemented is by looking at the gradient outputs at the beginning of backwards. If the gradient output corresponding to a model output is zero it means that it was not involved in loss computation (it could also mean that the gradient was indeed all zeros, but this is practically equivalent). In this case, we get the corresponding model output and mark that parameter as unused (which is equivalent to a zero gradient being accumulated).

Tested by ensuring that when a model output is not used in loss, the corresponding grad is not modified. Also verified that the grads are the same in local vs DDP training case.

Differential Revision: [D28048628](https://our.internmc.facebook.com/intern/diff/D28048628/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[DDP] Support not all outputs used in loss calculation"


Changes in this diff:
1. Enable passthrough autograd function for all cases, not just static.
2. With above, move `prepare_for_backward` which does unused parameter checking logic to beginning of backwards pass
3. Enhance process of unused parameter checking to account for outputs not being used in loss.

The way (3) is implemented is by looking at the gradient outputs at the beginning of backwards. If the gradient output corresponding to a model output is zero it means that it was not involved in loss computation (it could also mean that the gradient was indeed all zeros, but this is practically equivalent). In this case, we get the corresponding model output and mark that parameter as unused (which is equivalent to a zero gradient being accumulated).

Tested by ensuring that when a model output is not used in loss, the corresponding grad is not modified. Also verified that the grads are the same in local vs DDP training case.

Differential Revision: [D28048628](https://our.internmc.facebook.com/intern/diff/D28048628/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[DDP] Support not all outputs used in loss calculation"


Changes in this diff:
1. Enable passthrough autograd function for all cases, not just static.
2. With above, move `prepare_for_backward` which does unused parameter checking logic to beginning of backwards pass
3. Enhance process of unused parameter checking to account for outputs not being used in loss.

The way (3) is implemented is by looking at the gradient outputs at the beginning of backwards. If the gradient output corresponding to a model output is zero it means that it was not involved in loss computation (it could also mean that the gradient was indeed all zeros, but this is practically equivalent, since the gradient accumulated on the leaf will be zero by the chain rule). In this case, we get the corresponding model output and mark that parameter as unused (which is equivalent to a zero gradient being accumulated).

Tested by ensuring that when a model output is not used in loss, the corresponding grad is not modified. Also verified that the grads are the same in local vs DDP training case.

Differential Revision: [D28048628](https://our.internmc.facebook.com/intern/diff/D28048628/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[DDP] Support not all outputs used in loss calculation"


Changes in this diff:
1. Enable passthrough autograd function for all cases, not just static.
2. With above, move `prepare_for_backward` which does unused parameter checking logic to beginning of backwards pass
3. Enhance process of unused parameter checking to account for outputs not being used in loss.

The way (3) is implemented is by looking at the gradient outputs at the beginning of backwards. If the gradient output corresponding to a model output is zero it means that it was not involved in loss computation (it could also mean that the gradient was indeed all zeros, but this is practically equivalent, since the gradient accumulated on the leaf will be zero by the chain rule). In this case, we get the corresponding model output and mark that parameter as unused (which is equivalent to a zero gradient being accumulated).

Tested by ensuring that when a model output is not used in loss, the corresponding grad is not modified. Also verified that the grads are the same in local vs DDP training case.

Differential Revision: [D28048628](https://our.internmc.facebook.com/intern/diff/D28048628/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[DDP] Support not all outputs used in loss calculation"


Changes in this diff:
1. Enable passthrough autograd function when find_unused_parameters=True.
2. With above, move `prepare_for_backward` which does unused parameter checking logic to beginning of backwards pass, only when `find_unused_parameters=True`. 
3. Enhance process of unused parameter checking to account for outputs not being used in loss.

The way (3) is implemented is by looking at the gradient outputs at the beginning of backwards. If the gradient output corresponding to a model output is zero it means that it was not involved in loss computation (it could also mean that the gradient was indeed all zeros, but this is practically equivalent). In this case, we get the corresponding model output and mark that parameter as unused (which is equivalent to a zero gradient being accumulated) so that we don't touch the gradient.

Tested by ensuring that when a model output is not used in loss, the corresponding grad is not modified. Also verified that the grads are the same in local vs DDP training case.

Note that in this diff we are not enabling the pass through autograd function for regular case find_unused_parameters=False because that has a much bigger blast radius and needs additional careful analysis especially with regard to the performance. 


Differential Revision: [D28048628](https://our.internmc.facebook.com/intern/diff/D28048628/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[DDP] Support not all outputs used in loss calculation"


Changes in this diff:
1. Enable passthrough autograd function when find_unused_parameters=True.
2. With above, move `prepare_for_backward` which does unused parameter checking logic to beginning of backwards pass, only when `find_unused_parameters=True`. 
3. Enhance process of unused parameter checking to account for outputs not being used in loss.

The way (3) is implemented is by looking at the gradient outputs at the beginning of backwards. If the gradient output corresponding to a model output is zero it means that it was not involved in loss computation (it could also mean that the gradient was indeed all zeros, but this is practically equivalent). In this case, we get the corresponding model output and mark that parameter as unused (which is equivalent to a zero gradient being accumulated) so that we don't touch the gradient.

Tested by ensuring that when a model output is not used in loss, the corresponding grad is not modified. Also verified that the grads are the same in local vs DDP training case.

Note that in this diff we are not enabling the pass through autograd function for regular case find_unused_parameters=False because that has a much bigger blast radius and needs additional careful analysis especially with regard to the performance. 


Differential Revision: [D28048628](https://our.internmc.facebook.com/intern/diff/D28048628/)

[ghstack-poisoned]

zhaojuanmao commented on torch/nn/parallel/distributed.py:
```diff
@@ -800,33 +835,32 @@ def forward(self, *inputs, **kwargs):
             else:
                 output = self.module(*inputs, **kwargs)
 
+            disable = False
```
why is this flag needed?

rohan-varma replied:
Thanks for flagging, accidentally send debug code. Removing.

zhaojuanmao commented on torch/nn/parallel/distributed.py:
```diff
@@ -800,33 +835,32 @@ def forward(self, *inputs, **kwargs):
-        if self.static_graph and self.num_iterations == 1:
-            # Need to grab list of tensors from user output in order to pass
-            # to custom autograd function.
+        if not disable and (self.find_unused_parameters and not self.static_graph) or (self.static_graph and self.num_iterations == 1):
```
line too long

zhaojuanmao commented on torch/csrc/distributed/c10d/init.cpp:
```diff
@@ -401,6 +401,10 @@ An enum-like class for built-in communication hooks: ``ALLREDUCE`` and ``FP16_CO
           &::c10d::Reducer::save_thread_local_state,
           py::call_guard<py::gil_scoped_release>())
+      .def(
+          "set_per_iteration_param_outputs_unused",
```
nit: make it private

zhaojuanmao commented on torch/nn/parallel/distributed.py:
```diff
@@ -800,33 +835,32 @@ def forward(self, *inputs, **kwargs):
-                    self.reducer.prepare_for_backward([])
+                if disable or not self.find_unused_parameters or self.static_graph:
+                    if self.find_unused_parameters:
+                        self.reducer.prepare_for_backward(list(_find_tensors(output)))
```
in which case this will happen?

rohan-varma replied:
Yeah, with this change we should never need to call `self.reducer.prepare_for_backward(list(_find_tensors(output)))`. The suggestion below is much clearer.

zhaojuanmao commented on torch/nn/parallel/distributed.py:
```diff
@@ -800,33 +835,32 @@ def forward(self, *inputs, **kwargs):
-                    self.reducer.prepare_for_backward(list(_find_tensors(output)))
-                else:
-                    self.reducer.prepare_for_backward([])
+                if disable or not self.find_unused_parameters or self.static_graph:
+                    if self.find_unused_parameters:
+                        self.reducer.prepare_for_backward(list(_find_tensors(output)))
+                    else:
+                        self.reducer.prepare_for_backward([])
```
should it be just like this?

if self.static_graph or not self.find_unused_parameters:
    self.reducer.prepare_for_backward([])

rohan-varma replied:
Yep - this is much clearer. Thanks!

rohan-varma committed:
Update on "[DDP] Support not all outputs used in loss calculation"


Changes in this diff:
1. Enable passthrough autograd function when find_unused_parameters=True.
2. With above, move `prepare_for_backward` which does unused parameter checking logic to beginning of backwards pass, only when `find_unused_parameters=True`. 
3. Enhance process of unused parameter checking to account for outputs not being used in loss.

The way (3) is implemented is by triggering the autograd hook corresponding to parameters that did not participate in loss computation. Since they did not participate, the autograd hook is triggered with a gradient of zero, and if the gradient output corresponding to a model output is zero it means that it was not involved in loss computation (it could also mean that the gradient was indeed all zeros, but this is practically equivalent). 
Tested by ensuring that when a model output is not used in loss, the corresponding grad is not modified. Also verified that the grads are the same in local vs DDP training case.

Note that in this diff we are not enabling the pass through autograd function for regular case find_unused_parameters=False because that has a much bigger blast radius and needs additional careful analysis especially with regard to the performance. 


Differential Revision: [D28048628](https://our.internmc.facebook.com/intern/diff/D28048628/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[DDP] Support not all outputs used in loss calculation"


Changes in this diff:
1. Enable passthrough autograd function when find_unused_parameters=True.
2. With above, move `prepare_for_backward` which does unused parameter checking logic to beginning of backwards pass, only when `find_unused_parameters=True`. 
3. Enhance process of unused parameter checking to account for outputs not being used in loss.

The way (3) is implemented is by triggering the autograd hook corresponding to parameters that did not participate in loss computation. Since they did not participate, the autograd hook is triggered with a gradient of zero, and if the gradient output corresponding to a model output is zero it means that it was not involved in loss computation (it could also mean that the gradient was indeed all zeros, but this is practically equivalent). 
Tested by ensuring that when a model output is not used in loss, the corresponding grad is not modified. Also verified that the grads are the same in local vs DDP training case.

Note that in this diff we are not enabling the pass through autograd function for regular case find_unused_parameters=False because that has a much bigger blast radius and needs additional careful analysis especially with regard to the performance. 


Differential Revision: [D28048628](https://our.internmc.facebook.com/intern/diff/D28048628/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[DDP] Support not all outputs used in loss calculation"


Changes in this diff:
1. Enable passthrough autograd function when find_unused_parameters=True.
2. With above, move `prepare_for_backward` which does unused parameter checking logic to beginning of backwards pass, only when `find_unused_parameters=True`. 
3. Enhance process of unused parameter checking to account for outputs not being used in loss.

The way (3) is implemented is by triggering the autograd hook corresponding to parameters that did not participate in loss computation. Since they did not participate, the autograd hook is triggered with a gradient of zero, and if the gradient output corresponding to a model output is zero it means that it was not involved in loss computation (it could also mean that the gradient was indeed all zeros, but this is practically equivalent). 
Tested by ensuring that when a model output is not used in loss, the corresponding grad is not modified. Also verified that the grads are the same in local vs DDP training case.

Note that in this diff we are not enabling the pass through autograd function for regular case find_unused_parameters=False because that has a much bigger blast radius and needs additional careful analysis especially with regard to the performance. 


Differential Revision: [D28048628](https://our.internmc.facebook.com/intern/diff/D28048628/)

[ghstack-poisoned]

zhaojuanmao commented on torch/nn/parallel/distributed.py:
```diff
@@ -113,14 +113,31 @@ class _DDPUnevenInputsConfig(NamedTuple):
+            # for this call to backward. Due to this passthrough autograd
+            # function, autograd hooks for these parameters are now triggered
+            # with zero gradient to maintain parity with local training.
+            ctx.reducer.prepare_for_backward(list(_find_tensors(ctx.inputs)))
```
does it mean the unused parameters due to unused output will have zero gradients instead of being 'undefined'? I'm not sure whether this is OK?
also I'm curious about why gradients of unused output tensors are not None..?

rohan-varma replied:
@zhaojuanmao Yes, if a parameter goes unused because it wasn't used as part of loss computation it will have a zero gradient. This is what the autograd custom function does even on a local model, if it does not receive a grad for the tensor in the backwards pass, it fills it in with zero, see: 

Although, IIUC this shouldn't make a difference? If the grad for the param is k, then we accumulate a grad of `0` onto it and the grad is the same. On the other hand, if grad is `None`, then it will be changed to `0`, but this would happen with `model.zero_grad()` anyways?

rohan-varma replied:
@zhaojuanmao This should be fixed now with the above ` ctx.set_materialize_grads(False)`.

rohan-varma committed:
Update on "[DDP] Support not all outputs used in loss calculation"


Changes in this diff:
1. Enable passthrough autograd function when find_unused_parameters=True.
2. With above, move `prepare_for_backward` which does unused parameter checking logic to beginning of backwards pass, only when `find_unused_parameters=True`. 
3. Enhance process of unused parameter checking to account for outputs not being used in loss.

The way (3) is implemented is by triggering the autograd hook corresponding to parameters that did not participate in loss computation. Since they did not participate, the autograd hook is triggered with a gradient of zero, and if the gradient output corresponding to a model output is zero it means that it was not involved in loss computation (it could also mean that the gradient was indeed all zeros, but this is practically equivalent). 
Tested by ensuring that when a model output is not used in loss, the corresponding grad is not modified. Also verified that the grads are the same in local vs DDP training case.

Note that in this diff we are not enabling the pass through autograd function for regular case find_unused_parameters=False because that has a much bigger blast radius and needs additional careful analysis especially with regard to the performance. 


Differential Revision: [D28048628](https://our.internmc.facebook.com/intern/diff/D28048628/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[DDP] Support not all outputs used in loss calculation"


Changes in this diff:
1. Enable passthrough autograd function when find_unused_parameters=True.
2. With above, move `prepare_for_backward` which does unused parameter checking logic to beginning of backwards pass, only when `find_unused_parameters=True`. 
3. Enhance process of unused parameter checking to account for outputs not being used in loss.

The way (3) is implemented is by triggering the autograd hook corresponding to parameters that did not participate in loss computation. Since they did not participate, the autograd hook is triggered with a gradient of zero, and if the gradient output corresponding to a model output is zero it means that it was not involved in loss computation (it could also mean that the gradient was indeed all zeros, but this is practically equivalent). 
Tested by ensuring that when a model output is not used in loss, the corresponding grad is not modified. Also verified that the grads are the same in local vs DDP training case.

Note that in this diff we are not enabling the pass through autograd function for regular case find_unused_parameters=False because that has a much bigger blast radius and needs additional careful analysis especially with regard to the performance. 


Differential Revision: [D28048628](https://our.internmc.facebook.com/intern/diff/D28048628/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[DDP] Support not all outputs used in loss calculation"


Changes in this diff:
1. Enable passthrough autograd function when find_unused_parameters=True.
2. With above, move `prepare_for_backward` which does unused parameter checking logic to beginning of backwards pass, only when `find_unused_parameters=True`. 
3. Enhance process of unused parameter checking to account for outputs not being used in loss.

The way (3) is implemented is by triggering the autograd hook corresponding to parameters that did not participate in loss computation. Since they did not participate, the autograd hook is triggered with a gradient of None, and the reducer handles this appropriately to ensure that the gradient is not touched.

Tested by ensuring that when a model output is not used in loss, the corresponding grad is not modified. Also verified that the grads are the same in local vs DDP training case. Also verified that gradients are not touched in this case, i.e. if grad is originally `None`, it stays as `None`, not zero, after.

Note that in this diff we are not enabling the pass through autograd function for regular case find_unused_parameters=False because that has a much bigger blast radius and needs additional careful analysis especially with regard to the performance. 


Differential Revision: [D28048628](https://our.internmc.facebook.com/intern/diff/D28048628/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[DDP] Support not all outputs used in loss calculation"


Changes in this diff:
1. Enable passthrough autograd function when find_unused_parameters=True.
2. With above, move `prepare_for_backward` which does unused parameter checking logic to beginning of backwards pass, only when `find_unused_parameters=True`. 
3. Enhance process of unused parameter checking to account for outputs not being used in loss.

The way (3) is implemented is by triggering the autograd hook corresponding to parameters that did not participate in loss computation. Since they did not participate, the autograd hook is triggered with a gradient of None, and the reducer handles this appropriately to ensure that the gradient is not touched.

Tested by ensuring that when a model output is not used in loss, the corresponding grad is not modified. Also verified that the grads are the same in local vs DDP training case. Also verified that gradients are not touched in this case, i.e. if grad is originally `None`, it stays as `None`, not zero, after.

Note that in this diff we are not enabling the pass through autograd function for regular case find_unused_parameters=False because that has a much bigger blast radius and needs additional careful analysis especially with regard to the performance. 


Differential Revision: [D28048628](https://our.internmc.facebook.com/intern/diff/D28048628/)

[ghstack-poisoned]

zhaojuanmao commented on torch/lib/c10d/reducer.cpp:
```diff
@@ -526,7 +536,18 @@ void Reducer::autograd_hook(VariableIndex index) {
+    // globally, we don't want to touch the .grad field of the param.
+
+    auto& variable = get_param_from_index(index);
+    runGradCallbackForVariable(variable, [&](auto& grad) {
```
why the runGradCallbackForVariable is needed here? grad is not mutated anyway

rohan-varma replied:
We have to access `variable.grad()` right? When running in just DDP, this can be done with `variable.grad()` directly, but I believe this callback is needed when running in RPC context, where the grad is not available in variable.grad() field.

zhaojuanmao replied:
a, you are right!

zhaojuanmao reviewed (approved):
lgtm, thanks!

rohan-varma committed:
Update on "[DDP] Support not all outputs used in loss calculation"


Changes in this diff:
1. Enable passthrough autograd function when find_unused_parameters=True.
2. With above, move `prepare_for_backward` which does unused parameter checking logic to beginning of backwards pass, only when `find_unused_parameters=True`. 
3. Enhance process of unused parameter checking to account for outputs not being used in loss.

The way (3) is implemented is by triggering the autograd hook corresponding to parameters that did not participate in loss computation. Since they did not participate, the autograd hook is triggered with a gradient of None, and the reducer handles this appropriately to ensure that the gradient is not touched.

Tested by ensuring that when a model output is not used in loss, the corresponding grad is not modified. Also verified that the grads are the same in local vs DDP training case. Also verified that gradients are not touched in this case, i.e. if grad is originally `None`, it stays as `None`, not zero, after.

Note that in this diff we are not enabling the pass through autograd function for regular case find_unused_parameters=False because that has a much bigger blast radius and needs additional careful analysis especially with regard to the performance. 


Differential Revision: [D28048628](https://our.internmc.facebook.com/intern/diff/D28048628/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[DDP] Support not all outputs used in loss calculation"


Changes in this diff:
1. Enable passthrough autograd function when find_unused_parameters=True.
2. With above, move `prepare_for_backward` which does unused parameter checking logic to beginning of backwards pass, only when `find_unused_parameters=True`. 
3. Enhance process of unused parameter checking to account for outputs not being used in loss.

The way (3) is implemented is by triggering the autograd hook corresponding to parameters that did not participate in loss computation. Since they did not participate, the autograd hook is triggered with a gradient of None, and the reducer handles this appropriately to ensure that the gradient is not touched.

Tested by ensuring that when a model output is not used in loss, the corresponding grad is not modified. Also verified that the grads are the same in local vs DDP training case. Also verified that gradients are not touched in this case, i.e. if grad is originally `None`, it stays as `None`, not zero, after.

Note that in this diff we are not enabling the pass through autograd function for regular case find_unused_parameters=False because that has a much bigger blast radius and needs additional careful analysis especially with regard to the performance. 


Differential Revision: [D28048628](https://our.internmc.facebook.com/intern/diff/D28048628/)

[ghstack-poisoned]
